### DISCUSSION TITLE
Emacs as TCP server

### DISCUSSION THREAD

**Discussion by rslack** [joaotavora/eglot#1227] [Q&A]
Is there some way for eglot to have Emacs be the TCP server, allowing the language server connect to it?
This is the only connection method the language server I am working with supports, and it works for both vscode and lsp-mode.

Perhaps this falls under the idea of an " experimental LSP server" for eglot, and I need to subclass eglot-lsp-server and configure CONTACT with cons (CLASS-NAME . INITARGS)? It does not seem like `jsonrpc-process-connection`  supports this method either though.

**Comment by joaotavora** [comment]
> Is there some way for eglot to have Emacs be the TCP server, allowing the language server connect to it?

No, not currently. EDIT: or rather, not "out of the box".

>  It does not seem like jsonrpc-process-connection supports this method either though.

Though JSONRPC and `jsonrpc.el` is agnostic to the server/client roles (thus it talks about "endpoints" instead) the underlying transports need to have well-defined roles.  Nevertheless it is quite possible to make a JSONRPC server in `jsonrpc.el` and `jsonrpc-process-connection`.   You can check `test/lisp/jsonrpc-tests.el` for an example (perhaps not super-easy to read, but an example nonetheless).

But that's bare JSONRPC.  Eglot's _use_ of `jsonrpc-process-connection` was designed and primarily intended for connections made with `make-process` or `open-network-stream` which return a "process" object that [1] takes the role of the client. That's because initiating a connection almost always comes from an interactive `M-x eglot` invocation and it makes sense that the initiating endpoint is a TCP client (with an ephemeral port).

I can't very well see how having Eglot as a server can work.  Well, maybe more or less.  Do you envision `M-x eglot` to temporarily open a server listening for connections and then return the first one (and stopping the listen)?  

If so, it should theoretically be possible by making a sophisticated `CONTACT` _function_ (no need for the class thing, in principle) that starts this server, blockingly waits for connections and then directly returns a plist looking like `(:process FIRST-CONNECTED-CLIENT-PROCESS-HERE)`. 

The plist return value for the `CONTACT` function isn't documented, but reading `eglot-connect`  it should work.

_If_ that works, the `(CLASS-NAME . INITARGS)` could also work, of course, as long as you make `CONTACT` a function that does this crucial step of opening the server, waiting some time for a client, and then putting this client in the return value.  

HTH



Perhaps this should have been reflected in the name, but alas it wasn't.

[1]: in Emacs parlance any connection to outside the Emacs OS process is another "process"

**Reply by joaotavora** [reply]
Maybe start playing around with this function to understand how it works.  Then `test` might become the CONTACT:
```lisp
(defun test ()
  (let ((retval
         (catch 'done
           (let (timer server)
             (unwind-protect
                 (progn ; 10 second timeout
                   (setq timer (run-at-time 10 nil (lambda () (throw 'done :noone))))
                   (setq server (make-network-process
                                 :name "Eglot listener"
                                 :server t :host "localhost"
                                 :service 4444
                                 :log (lambda (listen-server client _message)
                                        (cancel-timer timer)
                                        (throw 'done `(:process ,client)))))
                   (while t (accept-process-output nil 30))) ; standard boilerplate for blocking
               (cancel-timer timer)
               (if (processp server) (delete-process server)))))))
    (when (eq retval :noone) (user-error "Noone connected!"))
    retval))
```
